homeassistant/components/homee/entity.py

```diff
@@ -0,0 +1,147 @@
+
+    @property
+    def available(self) -> bool:
+        """Return the availability of the underlying node."""
+        return self._node.state <= 1
+
+    @property
+    def should_poll(self) -> bool:
+        """Return if the entity should poll."""
+        return False
+
+    @property
+    def raw_data(self):
+        """Return the raw data of the node."""
+        return self._node.raw_data
+
+    @property
+    def extra_state_attributes(self) -> dict[str, dict[str, Any]] | None:
+        """Return entity specific state attributes."""
+        data = {}
+
+        if self._entry.options.get(CONF_ADD_HOMEE_DATA, False):
+            data[ATTR_HOMEE_DATA] = self._homee_data
+
+        return data if data else None
+
+    async def async_update(self):
+        """Fetch new state data for this node."""
+        homee = self._entry.runtime_data
+        await homee.update_node(self._node.id)
+
+    def register_listener(self):
+        """Register the on_changed listener on the node."""
+        self._clear_node_listener = self._node.add_on_changed_listener(
+            self._on_node_updated
+        )
+
+    def clear_listener(self):
+        """Clear the on_changed listener on the node."""
+        if self._clear_node_listener is not None:
+            self._clear_node_listener()
+
+    def attribute(self, attribute_type):
+        """Try to get the current value of the attribute of the given type."""
+        try:
+            attribute = self._node.get_attribute_by_type(attribute_type)
+        except KeyError:
+            raise AttributeNotFoundException(attribute_type) from None
+
+        # If the unit of the attribute is 'text', it is stored in .data
+        if attribute.unit == "text":
+            return self._node.get_attribute_by_type(attribute_type).data
+
+        return self._node.get_attribute_by_type(attribute_type).current_value
+
+    def _get_software_version(self) -> str:
+        """Return the software version of the node."""
+        if self.has_attribute(AttributeType.FIRMWARE_REVISION):
+            sw_version = self.attribute(AttributeType.FIRMWARE_REVISION)
+        elif self.has_attribute(AttributeType.SOFTWARE_REVISION):
+            sw_version = self.attribute(AttributeType.SOFTWARE_REVISION)
+        else:
+            sw_version = "undefined"
+
+        return sw_version
+
+    def get_attribute(self, attribute_type):
+        """Get the attribute object of the given type."""
+        return self._node.get_attribute_by_type(attribute_type)
+
+    def has_attribute(self, attribute_type):
+        """Check if an attribute of the given type exists."""
+        return attribute_type in self._node.attribute_map
+
+    def is_reversed(self, attribute_type) -> bool:
+        """Check if movement direction is reversed."""
+        attribute = self._node.get_attribute_by_type(attribute_type)
+        if hasattr(attribute.options, "reverse_control_ui"):
```

Comment:
why the hassattr?

Reply:
To check if it is there before I try to read it, because else I get a´n exception.
Is there another way?

Reply:
But I'm wondering why it wouldn't exist?

Reply:
Because only few devices have it. And if not, it is not present.
I don't know, if it can be "False" if it is present.

Reply:
Oh you mean you have different sub classes and some have this and some don't?

Reply:
OK, to shed some light on this, this is how such an attribute looks like:
```
{
            "id": 138,
            "node_id": 3,
            "instance": 0,
            "minimum": 0,
            "maximum": 4,
            "current_value": 1.0,
            "target_value": 1.0,
            "last_value": 4.0,
            "unit": "n%2Fa",
            "step_value": 1.0,
            "editable": 1,
            "type": 135,
            "state": 1,
            "last_changed": 1687175680,
            "changed_by": 1,
            "changed_by_id": 0,
            "based_on": 1,
            "data": "",
            "name": "",
            "options": {
                "can_observe": [300],
                "observes": [75],
                "reverse_control_ui": true,
                "automations": ["toggle"]
            }
        }
```
At the end you can see the `options` part. Every key:value pair therein is optional and may not be present.

Reply:
is this explained and ok then?